extensions/binding/org.eclipse.smarthome.binding.mqtt.generic/src/main/java/org/eclipse/smarthome/binding/mqtt/generic/handler/AbstractThingHandler.java

```diff
@@ -0,0 +1,157 @@
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Base handler
+ *
+ * @author David Graeff - Initial contribution
+ */
+@NonNullByDefault
+public abstract class AbstractThingHandler extends BaseThingHandler implements ChannelStateUpdateListener {
+    private final Logger logger = LoggerFactory.getLogger(AbstractThingHandler.class);
+    protected final @Nullable TransformationServiceProvider transformationServiceProvider;
+    protected final MqttChannelTypeProvider provider;
+
+    protected @Nullable MqttBrokerConnection connection;
+
+    public AbstractThingHandler(Thing thing, MqttChannelTypeProvider provider,
+            @Nullable TransformationServiceProvider transformationServiceProvider) {
+        super(thing);
+        this.provider = provider;
+        this.transformationServiceProvider = transformationServiceProvider;
+    }
+
+    /**
+     * Return the channel state for the given channelUID.
+     *
+     * @param channelUID The channelUID
+     * @return A channel state. May be null.
+     */
+    abstract protected @Nullable ChannelState getChannelState(ChannelUID channelUID);
+
+    /**
+     * Start the topic discovery and subscribe to all channel state topics on all {@link ChannelState}s.
+     * Put the thing ONLINE or OFFLINE.
+     *
+     * @param connection A started broker connection
+     */
+    abstract protected void start(MqttBrokerConnection connection) throws MqttException;
+
+    @Override
+    public void handleCommand(ChannelUID channelUID, Command command) {
+        if (connection == null) {
+            return;
+        }
+
+        final @Nullable ChannelState data = getChannelState(channelUID);
+
+        if (data == null) {
+            logger.warn("Channel {} not supported", channelUID.getId());
+            return;
+        }
+
+        if (command instanceof RefreshType || data.isReadOnly()) {
+            updateState(channelUID.getId(), data.getValue().getValue());
+            return;
+        }
+
+        data.setValue(command)
+                .thenRun(() -> logger.debug("Successfully published value {} to topic {}", command, data.getTopic()))
+                .exceptionally(e -> {
+                    updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR, e.getLocalizedMessage());
+                    return null;
+                });
+    }
+
+    @Override
+    public void bridgeStatusChanged(ThingStatusInfo bridgeStatusInfo) {
+        if (bridgeStatusInfo.getStatus() == ThingStatus.OFFLINE) {
+            updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.BRIDGE_OFFLINE);
+            connection = null;
+            return;
+        }
+        if (bridgeStatusInfo.getStatus() != ThingStatus.ONLINE) {
+            return;
+        }
+
+        AbstractBrokerHandler h = getBridgeHandler();
+        if (h == null) {
+            return;
+        }
+        connection = h.getConnection();
+        MqttBrokerConnection connection = this.connection;
```

Comment:
It is to avoid null warnings. I will change the code to:
```java
this.connection = h.getConnection();
MqttBrokerConnection connection = this.connection;
```
to make it more clear.